### PR TITLE
Addition of Filestore API

### PR DESCRIPTION
**Pull request by nguyenv**

### CODE REVIEW

**Comment by nguyenv** [comment]
https://github.com/TileDB-Inc/TileDB/pull/3121

**Review by ihnorton — COMMENTED** [review]
There's an uncaught exception somewhere - if I run the following (without creating a schema + array first):
```
#%%
import tiledb
# %%
fs = tiledb.Filestore("test_arrays/GitUp.zip")
fs.copy_from("test_files/GitUp.zip")
```

get:
```
libc++abi.dylib: terminating with uncaught exception of type tiledb::TileDBError: [TileDB::Array] Error: Cannot open array; Array does not exist
```

File: tiledb/filestore.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+            )
+
+        lt.Filestore._uri_import(
+            lt.Context(self._ctx.__capsule__(), False),
+            self._filestore_uri,
+            uri,
+            lt.Filestore._mime_type_from_str(mime_type),
+        )
+
+    def copy_to(self, uri: str) -> None:
```
Comment (ihnorton) [inline tiledb/filestore.py:97]: Not immediately necessary, but would be nice if `copy_to` checked if the output is a path and inferred the full URI based on the stored filename.

**Comment by nguyenv** [comment]
> There's an uncaught exception somewhere - if I run the following (without creating a schema + array first):
> 
> ```
> #%%
> import tiledb
> # %%
> fs = tiledb.Filestore("test_arrays/GitUp.zip")
> fs.copy_from("test_files/GitUp.zip")
> ```
> 
> get:
> 
> ```
> libc++abi.dylib: terminating with uncaught exception of type tiledb::TileDBError: [TileDB::Array] Error: Cannot open array; Array does not exist
> ```

As mentioned on Slack, I get a different error when I attempt to call any of these functions (including `copy_from` like in your example), but the array doesn't exist.
```
(tiledb-3.10) vivian@mangonada:~/TileDB-Py$ pytest tiledb/tests/test_filestore.py
============================================================================================================ test session starts ============================================================================================================
platform linux -- Python 3.10.4, pytest-7.1.2, pluggy-1.0.0
rootdir: /home/vivian/TileDB-Py, configfile: pyproject.toml
plugins: hypothesis-6.45.1
collected 3 items

tiledb/tests/test_filestore.py Fatal Python error: Aborted

Current thread 0x00007f6b09549340 (most recent call first):
  File "/home/vivian/TileDB-Py/tiledb/filestore.py", line 43 in write
  File "/home/vivian/TileDB-Py/tiledb/tests/test_filestore.py", line 25 in test_buffer
  File "/home/vivian/anaconda3/envs/tiledb-3.10/lib/python3.10/site-packages/_pytest/python.py", line 192 in pytest_pyfunc_call
  File "/home/vivian/anaconda3/envs/tiledb-3.10/lib/python3.10/site-packages/pluggy/_callers.py", line 39 in _multicall
  File "/home/vivian/anaconda3/envs/tiledb-3.10/lib/python3.10/site-packages/pluggy/_manager.py", line 80 in _hookexec
  File "/home/vivian/anaconda3/envs/tiledb-3.10/lib/python3.10/site-packages/pluggy/_hooks.py", line 265 in __call__
  File "/home/vivian/anaconda3/envs/tiledb-3.10/lib/python3.10/site-packages/_pytest/python.py", line 1761 in runtest
  File "/home/vivian/anaconda3/envs/tiledb-3.10/lib/python3.10/site-packages/_pytest/runner.py", line 166 in pytest_runtest_call
  File "/home/vivian/anaconda3/envs/tiledb-3.10/lib/python3.10/site-packages/pluggy/_callers.py", line 39 in _multicall
  File "/home/vivian/anaconda3/envs/tiledb-3.10/lib/python3.10/site-packages/pluggy/_manager.py", line 80 in _hookexec
  File "/home/vivian/anaconda3/envs/tiledb-3.10/lib/python3.10/site-packages/pluggy/_hooks.py", line 265 in __call__
  File "/home/vivian/anaconda3/envs/tiledb-3.10/lib/python3.10/site-packages/_pytest/runner.py", line 259 in <lambda>
  File "/home/vivian/anaconda3/envs/tiledb-3.10/lib/python3.10/site-packages/_pytest/runner.py", line 338 in from_call
  File "/home/vivian/anaconda3/envs/tiledb-3.10/lib/python3.10/site-packages/_pytest/runner.py", line 258 in call_runtest_hook
  File "/home/vivian/anaconda3/envs/tiledb-3.10/lib/python3.10/site-packages/_pytest/runner.py", line 219 in call_and_report
  File "/home/vivian/anaconda3/envs/tiledb-3.10/lib/python3.10/site-packages/_pytest/runner.py", line 130 in runtestprotocol
  File "/home/vivian/anaconda3/envs/tiledb-3.10/lib/python3.10/site-packages/_pytest/runner.py", line 111 in pytest_runtest_protocol
  File "/home/vivian/anaconda3/envs/tiledb-3.10/lib/python3.10/site-packages/pluggy/_callers.py", line 39 in _multicall
  File "/home/vivian/anaconda3/envs/tiledb-3.10/lib/python3.10/site-packages/pluggy/_manager.py", line 80 in _hookexec
  File "/home/vivian/anaconda3/envs/tiledb-3.10/lib/python3.10/site-packages/pluggy/_hooks.py", line 265 in __call__
  File "/home/vivian/anaconda3/envs/tiledb-3.10/lib/python3.10/site-packages/_pytest/main.py", line 347 in pytest_runtestloop
  File "/home/vivian/anaconda3/envs/tiledb-3.10/lib/python3.10/site-packages/pluggy/_callers.py", line 39 in _multicall
  File "/home/vivian/anaconda3/envs/tiledb-3.10/lib/python3.10/site-packages/pluggy/_manager.py", line 80 in _hookexec
  File "/home/vivian/anaconda3/envs/tiledb-3.10/lib/python3.10/site-packages/pluggy/_hooks.py", line 265 in __call__
  File "/home/vivian/anaconda3/envs/tiledb-3.10/lib/python3.10/site-packages/_pytest/main.py", line 322 in _main
  File "/home/vivian/anaconda3/envs/tiledb-3.10/lib/python3.10/site-packages/_pytest/main.py", line 268 in wrap_session
  File "/home/vivian/anaconda3/envs/tiledb-3.10/lib/python3.10/site-packages/_pytest/main.py", line 315 in pytest_cmdline_main
  File "/home/vivian/anaconda3/envs/tiledb-3.10/lib/python3.10/site-packages/pluggy/_callers.py", line 39 in _multicall
  File "/home/vivian/anaconda3/envs/tiledb-3.10/lib/python3.10/site-packages/pluggy/_manager.py", line 80 in _hookexec
  File "/home/vivian/anaconda3/envs/tiledb-3.10/lib/python3.10/site-packages/pluggy/_hooks.py", line 265 in __call__
  File "/home/vivian/anaconda3/envs/tiledb-3.10/lib/python3.10/site-packages/_pytest/config/__init__.py", line 164 in main
  File "/home/vivian/anaconda3/envs/tiledb-3.10/lib/python3.10/site-packages/_pytest/config/__init__.py", line 187 in console_main
  File "/home/vivian/anaconda3/envs/tiledb-3.10/bin/pytest", line 8 in <module>

Extension modules: numpy.core._multiarray_umath, numpy.core._multiarray_tests, numpy.linalg._umath_linalg, numpy.fft._pocketfft_internal, numpy.random._common, numpy.random.bit_generator, numpy.random._bounded_integers, numpy.random._mt19937, numpy.random.mtrand, numpy.random._philox, numpy.random._pcg64, numpy.random._sfc64, numpy.random._generator, tiledb.libtiledb, pyarrow.lib, pyarrow._hdfsio, pandas._libs.tslibs.dtypes, pandas._libs.tslibs.base, pandas._libs.tslibs.np_datetime, pandas._libs.tslibs.nattype, pandas._libs.tslibs.timezones, pandas._libs.tslibs.ccalendar, pandas._libs.tslibs.tzconversion, pandas._libs.tslibs.strptime, pandas._libs.tslibs.fields, pandas._libs.tslibs.timedeltas, pandas._libs.tslibs.timestamps, pandas._libs.properties, pandas._libs.tslibs.offsets, pandas._libs.tslibs.parsing, pandas._libs.tslibs.conversion, pandas._libs.tslibs.period, pandas._libs.tslibs.vectorized, pandas._libs.ops_dispatch, pandas._libs.missing, pandas._libs.hashtable, pandas._libs.algos, pandas._libs.interval, pandas._libs.tslib, pandas._libs.lib, pandas._libs.hashing, pandas._libs.ops, pandas._libs.arrays, pandas._libs.index, pandas._libs.join, pandas._libs.sparse, pyarrow._compute, pandas._libs.reduction, pandas._libs.indexing, pandas._libs.internals, pandas._libs.writers, pandas._libs.window.aggregations, pandas._libs.window.indexers, pandas._libs.reshape, pandas._libs.groupby, pandas._libs.testing, pandas._libs.parsers, pandas._libs.json, numpy.linalg.lapack_lite (total: 59)
Aborted
```

GDB backtrace.
```
Thread 1 "python" received signal SIGABRT, Aborted.
__GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
50      ../sysdeps/unix/sysv/linux/raise.c: No such file or directory.
(gdb) bt
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
#1  0x00007ffff7c72859 in __GI_abort () at abort.c:79
#2  0x00007ffff4f85951 in ?? () from /lib/x86_64-linux-gnu/libstdc++.so.6
#3  0x00007ffff4f9147c in ?? () from /lib/x86_64-linux-gnu/libstdc++.so.6
#4  0x00007ffff4f90459 in ?? () from /lib/x86_64-linux-gnu/libstdc++.so.6
#5  0x00007ffff4f90e11 in __gxx_personality_v0 () from /lib/x86_64-linux-gnu/libstdc++.so.6
#6  0x00007ffff4edcbef in ?? () from /lib/x86_64-linux-gnu/libgcc_s.so.1
#7  0x00007ffff4edd5aa in _Unwind_Resume () from /lib/x86_64-linux-gnu/libgcc_s.so.1
#8  0x00007ffff59282fa in tiledb::Array::Array (this=0x7fffffff78a0, ctx=..., array_uri="/tmp/tiledb-filestoretestuxttbuxq/test_buffer", query_type=TILEDB_WRITE, encryption_type=TILEDB_NO_ENCRYPTION, encryption_key=0x0, key_length=0,
    timestamp=1651862239578) at /home/vivian/TileDB/tiledb/../tiledb/sm/cpp_api/array.h:259
#9  0x00007ffff59280ef in tiledb::Array::Array (this=0x7fffffff78a0, ctx=..., array_uri="/tmp/tiledb-filestoretestuxttbuxq/test_buffer", query_type=TILEDB_WRITE, timestamp=1651862239578)
    at /home/vivian/TileDB/tiledb/../tiledb/sm/cpp_api/array.h:218
#10 0x00007ffff592285d in tiledb::common::detail::tiledb_filestore_buffer_import (ctx=0x555556310370, filestore_array_uri=0x55555670fc30 "/tmp/tiledb-filestoretestuxttbuxq/test_buffer", buf=0x7fffd9943710, size=6,
    mime_type=TILEDB_MIME_AUTODETECT) at /home/vivian/TileDB/tiledb/sm/c_api/tiledb_filestore.cc:391
```

File: tiledb/tests/test_filestore.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+        with self.assertRaises(tiledb.TileDBError):
+            fs.write(data)
+
+        schema = tiledb.ArraySchema.from_file()
+        tiledb.Array.create(path, schema)
+
+        assert schema.attr(0).name == "contents"
+        assert schema.attr(0).dtype == np.bytes_
+
+        # TODO blocked by SC-17479. would like to check that the attr is infact
```
Comment (Shelnutt2) [inline tiledb/tests/test_filestore.py:35]: Fixed by: https://github.com/TileDB-Inc/TileDB/pull/3250

File: tiledb/libtiledb.pyx [hunk 3]
```diff
@@ -1275,7 +1275,10 @@ cdef _numpy_dtype(tiledb_datatype_t tiledb_dtype, cell_size = 1):
     cdef base_dtype
     cdef uint32_t cell_val_num = cell_size
 
-    if cell_val_num == 1:
+    if tiledb_dtype == TILEDB_BLOB:
```
Comment (Shelnutt2) [inline tiledb/libtiledb.pyx:1278]: Can we pull the BLOB types out to another PR so we can have compatibility array created with this type?

**Comment by nguyenv** [comment]
Confirmed that the uncaught exceptions have been fixed by https://github.com/TileDB-Inc/TileDB/pull/3273.

**Review by ihnorton — APPROVED** [review]
🎉 LGTM
